apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -141,7 +141,23 @@ export function parsePnpmDependencyKey(dependencyName: string, dependencyKey: st
   }
 
   if (!semver.valid(parsedVersionPart)) {
-    return undefined;
+    // Test for urls:
+    // Examples:
+    //     github.com/abc/def/188ed64efd5218beda276e02f2277bf3a6b745b2
+    //     github.com.au/abc/def/188ed64efd5218beda276e02f2277bf3a6b745b2
+    //     bitbucket.com/abc/def/188ed64efd5218beda276e02f2277bf3a6b745b2
+    //     bitbucket.co.in/abc/def/188ed64efd5218beda276e02f2277bf3a6b745b2
+    if (/^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}\/([^\/\\]+\/?)*([^\/\\]+)$/i.test(dependencyKey)) {
+      try {
+        const dependencySpecifier: DependencySpecifier = new DependencySpecifier(dependencyName, dependencyKey);
```

Comment:
The `dependencyKey` here does not seem to be a valid NPM version specifier.  The `npm-package-arg` library is interpreting it as shorthand for `file:bitbucket.com\\abc\\def\\188ed64efd5218beda276e02f2277bf3a6b745b2`.  In order to parse it correctly you would need to add a prefix like:
```ts
new DependencySpecifier(dependencyName, "git:" + dependencyKey);
```

But I feel like we might be mixing up kinds-of-things.  The original **package.json** in your example started with:

**package.json**
```js
  "dependencies": {
    "xmlrpc": "chfritz/node-xmlrpc"
  },
```

Then PNPM serialized it as:

**pnpm-lock.yaml**
```yaml
  'file:projects/demoapp.tgz':
    dependencies:
      typescript: 3.7.5
      xmlrpc: github.com/chfritz/node-xmlrpc/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38
    dev: false
```

...which refers to:

```yaml
  github.com/chfritz/node-xmlrpc/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38:
    dependencies:
      sax: 1.2.4
      xmlbuilder: 8.2.2
    dev: false
    engines:
      node: '>=0.8'
      npm: '>=1.0.0'
    name: xmlrpc
    resolution:
      tarball: 'https://codeload.github.com/chfritz/node-xmlrpc/tar.gz/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38'
    version: 1.3.2
```

So the three formats are:
1. `chfritz/node-xmlrpc` - the original shorthand
2. `github.com/chfritz/node-xmlrpc/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38` - the shrinkwrap key
3. `https://codeload.github.com/chfritz/node-xmlrpc/tar.gz/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38` -- the actual download URL

I'm wondering whether form (2) is some kind of special pattern (i.e. that we need to match carefully), or if it's just a generic lookup key, and it's really (3) that really tells us that this is a Git URL (i.e. that we need to follow and compare instead).  I couldn't easily determine that from digging around in the PNPM source code.

Reply:
@zkochan What's the right way to determine that `xmlrpc: github.com/chfritz/node-xmlrpc/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38` is a Git specifier?

Reply:
I don't think you could determine with 100% confidence if a package was resolved through Git, using the lockfile. The lockfile doesn't really care how it was resolved. It has the information needed for fetching. In this case, fetching doesn't even need Git, because Github has an API for fetching tarballs.

Reply:
@octogonz I think it's okay to ignore `npm-package-arg` interpreting the specifier type incorrectly, at least for now.